Return predictions from all models in ensemble

cjmcgill opened the pull request:
Added a prediction argument `--individual_ensemble_predictions` that returns the predictions from each individual model in an ensemble in addition to the average ensemble prediction.

hesther commented on chemprop/train/make_predictions.py:
```diff
@@ -129,15 +129,15 @@ def make_predictions(args: PredictArgs, smiles: List[List[str]] = None) -> List[
     for full_index, datapoint in enumerate(full_data):
         valid_index = full_to_valid_indices.get(full_index, None)
-        preds = avg_preds[valid_index] if valid_index is not None else ['Invalid SMILES'] * len(task_names)
+        preds = avg_preds[valid_index] if valid_index is not None else ['Invalid SMILES'] * num_tasks
```
Is num_tasks and len(task_names) equivalent? I am especially thinking about edge_cases, where e.g. one column was excluded by chemprop because it didn't hold a single label.

cjmcgill replied:
I don't think I see the place where an empty column would lead to that changing. We skip empty rows but do we skip empty columns?

This is an excellent note. I thought that these were equivalent, and they are at the start. `args.num_tasks` is not actually a normal attribute, it's a property that is just `len(args.task_names)`. However, it looks like the num_targets variable is updated when it's a multiclass job (my other problem!) so I need to take a closer look at this.

hesther commented:
Just ran a bunch of tests. This works great for regression models (single and multi-task), and classification models (single and multi-task) but it fails for multiclass classification:

```
Traceback (most recent call last):
  File "predict.py", line 6, in <module>
    chemprop_predict()
  File "/Users/esther/Documents/Work/MIT/Data/test_chemprop/chemprop/chemprop/train/make_predictions.py", line 178, in chemprop_predict
    make_predictions(args=PredictArgs().parse_args())
  File "/Users/esther/Documents/Work/MIT/Data/test_chemprop/chemprop/chemprop/utils.py", line 557, in wrap
    result = func(*args, **kwargs)
  File "/Users/esther/Documents/Work/MIT/Data/test_chemprop/chemprop/chemprop/train/make_predictions.py", line 112, in make_predictions
    all_preds[:, :, index] = model_preds
ValueError: setting an array element with a sequence.
```

hesther commented:
For multiclass classification, if e.g. num_tasks is 1 and multiclass_num_classes is 3, then the prediction output has three values, so I guess that needs a special treatment

cjmcgill commented:
@hesther I fixed the indexing issue that was causing problems for this (and would cause issues for the ensemble variance too if we had checked that).

I also fixed how the multiclass results are saved, which appeared to be broken before. Previously they were being saved as a list inside one column of the csv. However the renaming of the task names indicated that each entry was intended to be given its own column. And under the previous error system, if you had more than one multiclass target, your target names in the column headers would be wrong. So that stealth bug is fixed now.

hesther reviewed (approved):
Tested for regression, classification and multiclass classification - works well! Also, it is great that the saving issues for multiclass classification are now fixed.
